src/bvals/boundary_conditions.cpp

```diff
@@ -23,248 +23,237 @@
+        Real reflect = q.VectorComponent(l) == X1DIR ? -1.0 : 1.0;
+        q(l, k, j, i) = reflect * q(l, k, j, 2 * ref - i + 1);
+      });
+}
+
+void ReflectInnerX2(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse) {
+  std::shared_ptr<MeshBlock> pmb = rc->GetBlockPointer();
+  auto bounds = coarse ? pmb->c_cellbounds : pmb->cellbounds;
+  int ref = bounds.GetBoundsJ(IndexDomain::interior).s;
+  auto q = rc->PackVariables(std::vector<MetadataFlag>{Metadata::FillGhost}, coarse);
+  auto nb = IndexRange{0, q.GetDim(4) - 1};
+  pmb->par_for_bndry(
+      "ReflectInnerX2", nb, IndexDomain::inner_x2, coarse,
+      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+        Real reflect = q.VectorComponent(l) == X2DIR ? -1.0 : 1.0;
+        q(l, k, j, i) = reflect * q(l, k, 2 * ref - j - 1, i);
+      });
+}
+
+void ReflectOuterX2(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse) {
+  std::shared_ptr<MeshBlock> pmb = rc->GetBlockPointer();
+  auto bounds = coarse ? pmb->c_cellbounds : pmb->cellbounds;
+  int ref = bounds.GetBoundsJ(IndexDomain::interior).e;
+  auto q = rc->PackVariables(std::vector<MetadataFlag>{Metadata::FillGhost}, coarse);
+  auto nb = IndexRange{0, q.GetDim(4) - 1};
+  pmb->par_for_bndry(
+      "ReflectOuterX2", nb, IndexDomain::outer_x2, coarse,
+      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+        Real reflect = q.VectorComponent(l) == X2DIR ? -1.0 : 1.0;
+        q(l, k, j, i) = reflect * q(l, k, 2 * ref - j + 1, i);
+      });
+}
+
+void ReflectInnerX3(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse) {
+  std::shared_ptr<MeshBlock> pmb = rc->GetBlockPointer();
+  auto bounds = coarse ? pmb->c_cellbounds : pmb->cellbounds;
+  int ref = bounds.GetBoundsK(IndexDomain::interior).s;
+  auto q = rc->PackVariables(std::vector<MetadataFlag>{Metadata::FillGhost}, coarse);
+  auto nb = IndexRange{0, q.GetDim(4) - 1};
+  pmb->par_for_bndry(
+      "ReflectInnerX3", nb, IndexDomain::inner_x3, coarse,
+      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+        Real reflect = q.VectorComponent(l) == X3DIR ? -1.0 : 1.0;
+        q(l, k, j, i) = reflect * q(l, 2 * ref - k - 1, j, i);
+      });
+}
+
+void ReflectOuterX3(std::shared_ptr<MeshBlockData<Real>> &rc, bool coarse) {
+  std::shared_ptr<MeshBlock> pmb = rc->GetBlockPointer();
+  auto bounds = coarse ? pmb->c_cellbounds : pmb->cellbounds;
+  int ref = bounds.GetBoundsK(IndexDomain::interior).e;
+  auto q = rc->PackVariables(std::vector<MetadataFlag>{Metadata::FillGhost}, coarse);
+  auto nb = IndexRange{0, q.GetDim(4) - 1};
+  pmb->par_for_bndry(
+      "ReflectOuterX3", nb, IndexDomain::outer_x3, coarse,
+      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+        Real reflect = q.VectorComponent(l) == X3DIR ? -1.0 : 1.0;
+        q(l, k, j, i) = reflect * q(l, 2 * ref - k + 1, j, i);
+      });
+}
```

Comment:
There is a lot of repetitive logic here, is there anyway some of it could be packaged into a local subfunction?

Reply:
Yeah I don't love this the way it currently is. But no easy solution came to mind. The problem is that the KOKKOS_LAMBDA is slightly different each time in a way that makes it difficult to generalize more.

Also there may be some utility in leaving these as-is as reference implementations for downstream codes.

Suggestions welcome.

Reply:
Oh, I see what you mean, ya, I should have paid more attention.